src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+
+
+def get_tool_header(version: str) -> dict:
+    return {
+        "vendor": "nexB Inc.",
+        "name": "scancode-toolkit",
+        "version": version
+    }
+
+
+def get_external_ref_from_key(package: dict, key: str) -> CycloneDxExternalRef:
+    url = package.get(key)
+    type = url_type_mapping[key]
+    if url is not None and type is not None:
+        return CycloneDxExternalRef(url=url, type=type)
+    return None
+
+
+def get_license_entry_from_license_expression(license_expression: str) \
+        -> CycloneDxLicenseEntry:
+    """query our list of known licenses to see if we can resolve
+    the license_expression to a single entry"""
+
+    # exit early if we don't have a valid license_expression
+    if license_expression is None:
+        return None
+
+    license = known_licenses.get(license_expression)
+    if license is not None:
+        # attempt to set either official OSI URL or any license text URL
+        url = license.osi_url
+        if license.text_urls:
+            url = license.text_urls[0]
+        lic = CycloneDxLicense(id=license.spdx_license_key,
+                               name=license.name,
+                               url=url)
+        return CycloneDxLicenseEntry(license=lic, expression=None)
+    else:
+        return CycloneDxLicenseEntry(license=None, expression=license_expression)
+
+
+url_pattern = re.compile(r"^(https?://)?[^\s/$.?#].[^\s]*$")
+
+
+def get_license_entry_from_declared_license(declared_license) \
+        -> CycloneDxLicenseEntry:
+    lic = CycloneDxLicense()
+    if isinstance(declared_license, str):
+        # if license key is in list of known spdx ids, set as id
+        if declared_license in spdx_ids:
+            lic.id = declared_license
+        # if we match this regex assume we are dealing with a URL
+        elif declared_license is not None and url_pattern.match(declared_license):
+            lic.url = declared_license
+        else:
+            lic.name = declared_license
+    # some declared_license entries are also expressed as dicts
+    elif isinstance(declared_license, dict):
+        lic_type = declared_license.get("type")
+        if lic_type in spdx_ids:
+            lic.id = lic_type
+        else:
+            lic.name = lic_type
+        lic.url = declared_license.get("url")
+    return CycloneDxLicenseEntry(license=lic, expression=None)
+
+
+def get_licenses(package: dict) -> List[CycloneDxLicenseEntry]:
+    """map all license data for a package entry to
+    a list of CycloneDX Licenses"""
+    # store previously encountered IDs to identify duplicates
+    seen_ids = set()
+
+    lic_expr = package.get("license_expression")
+    entry = get_license_entry_from_license_expression(lic_expr)
+    if entry is not None:
+        licenses = [entry]
+        if entry.license is not None:
+            seen_ids.add(entry.license.id)
+    else:
+        licenses = []
+
+    declared_license = package["declared_license"]
```

Comment:
Do not use both declared_license and license expression, as this is redundant.

Reply:
I have a question about this:

In one of the larger tests I ran the `license_expression` produced by scancode and the declared package license looked like this:

```json
 "purl": "pkg:npm/has@1.0.3",
      "licenses": [
        {
          "expression": "mit AND (mit AND (mit AND unknown))"
        },
        {
          "license": {
            "id": "MIT"
          }
        }
      ],
```

I know too little about your license detection engine to explain where this detection comes from, but in that case having the declared license alongside the expression can be helpful.  

The approach I am currently taking here is:
- add the license expression as a license entry
- only add declared license(s) if they differ from the license expression

Given the example above, is this a viable option?